File: src/python/algorithms/fuerzaBruta.py
import numpy as np
import time
import numpy as np
from algorithms.Core import *


def fuerzaBruta(m,n,k,datos,dataName):
    start = time.time()
    #creamos la grilla en x y en y para los datos 
    gridX:list = np.linspace(min(datos["x"]), max(datos["x"]), num=m, endpoint=True)
    gridY:list = np.linspace(min(datos["y"]), max(datos["y"]), num=n, endpoint=True)

    res = [[],[]]
        
    #esto lo hice porque la lista estba vacia para darles valores inciales porque 
    #sino en la priemra iteracion no podia iterar sobre valoers vacios por eso daba el lsit index out of range [1]
    for i in range(k):
        res[0].append(gridX[i])
        res[1].append(gridY[-1])

    

    bestError = float('inf')
        
    bestError = fuerzaBrutaRecursiva(gridX,gridY,[],[],res,bestError,k,datos)



    end = time.time()
    totalTime = (end - start) * 1000



    plot_puntos_y_linea(datos,res[0],res[1],m,n,"FuerzaBruta",bestError,totalTime,dataName)
  
    #devolvemos: mejor error, solucion en x, solucion en y y el tiempo total
    return np.round(bestError,decimals=2),res,np.round(totalTime,decimals=2)

def fuerzaBrutaRecursiva(gridX:list,gridY:list,xs:list,ys:list,res:list,bestError:float,k:int, datos):
    BIG_NUMBER = 1000000001
    
    # CASO BASE
    if(k==0):
        # CASOS INFACTIBLES

        # no empieza en el principio del grid
        if xs[0] != gridX[0]:
            return BIG_NUMBER
        # no termina en el fin del grid
        if xs[-1] != gridX[-1]:
            return BIG_NUMBER
        # va para atras y/o arriba, o el mismo punto
        for i in range(k-1):
            if(xs[i] >= xs[i+1]):
                return BIG_NUMBER
        
        error = errorBreakPoints(xs,ys,datos)
        if(error < bestError):
                res[0].clear()
                res[1].clear()
                res[0].extend(xs)
                res[1].extend(ys)
                bestError = error
        
        return bestError

    # PASO RECURSIVO

    for i in range(len(gridX)):
        for j in range(len(gridY)):
            xs.append(gridX[i])
            ys.append(gridY[j])
            error = fuerzaBrutaRecursiva(gridX,gridY,xs,ys,res,bestError,k-1,datos)
            if(error < bestError):
                bestError = error
            xs.pop()
            ys.pop()

    return bestError

File: src/cpp/FuerzaBruta.cpp
#include "include/FuerzaBruta.hpp"

float fuerzaBrutaRecursiva(
    vector<float> &gridX,
    vector<float> &gridY,
    vector<float> &xs,
    vector<float> &ys,
    vector<float> &res,
    float bestError_,
    int k,
    const json &datos)
{

    //CASOS BASE

    if( k == 0 ){
        if(xs[0] != gridX[0]){
            return 100000001;
        }
        if(xs[xs.size()-1] != gridX[gridX.size()-1]){
            return 100000001;
        }

        for(int i=0; i< xs.size()-2;i++){
            if(xs[i] > xs[i+1] ){
                return 100000001;
            }
        }

        float error = errorBreakPoint(xs,ys,datos);

        if(error < bestError_){
            /*std::cout<<"ERROR: "<<error;
            printVector(&ys, " - BEST YS: ");
            */
            res = ys;
            res.clear();
            for(int i=0;i < ys.size();i++){
                res.push_back(ys[i]);
            }
            
            


            return error;
        }
        return bestError_;
    }


    //PASO RECURSIVO
    float bestError = bestError_;
    float error;

    for(int i=0; i<gridX.size();i++){
        xs.push_back(gridX[i]);
        for(int j=0; j< gridY.size();j++){
            //TODO
            ys.push_back(gridY[j]);
            error = fuerzaBrutaRecursiva(gridX,gridY,xs,ys,res,bestError,k-1,datos);

            if(error < bestError){
                bestError = error;
            }
            
            ys.pop_back();

        }
        xs.pop_back();
    }

    return bestError;

}

void fuerzaBruta(Resultado &res, int m, int n,int k, const json &data){
    auto inicio = high_resolution_clock::now();

    std::vector<float> grillaX;
    std:vector<float> grillaY;

    linspace(grillaX,minJson(data,"x"),maxJson(data,"x"),m);
    linspace(grillaY, minJson(data, "y"), maxJson(data, "y"), n);

    vector<float> xs;
    vector<float> ys;
    vector<float> result;

    float bestError = fuerzaBrutaRecursiva(grillaX,grillaY,xs,ys,result,10000001,k,data);

    // res.bestError = bestError;
    
    auto fin = high_resolution_clock::now();
    double duracion = (double)duration_cast<duration<double>>(fin - inicio).count();
    

    std::cout << "BEST_ERROR: " << bestError << "  TIME: " << duracion << endl;
    res.time = duracion;
    res.bestError = bestError;

    printVector(&result);
}
